boards/qt_py_m0/examples/usb_echo.rs

```diff
@@ -0,0 +1,103 @@
+#![no_std]
+#![no_main]
+
+//! USB CDC example for the Adafruit QT Py board. Demonstrates creating a USB
+//! CDC ACM serial port accessible over the on-board USB-C connector.
+
+use panic_halt as _;
+
+use cortex_m::peripheral::NVIC;
+use nb::block;
+use usb_device::bus::UsbBusAllocator;
+use usb_device::prelude::*;
+use usbd_serial::SerialPort;
+use usbd_serial::USB_CLASS_CDC;
+
+use bsp::hal;
+use bsp::pac;
+use qt_py_m0 as bsp;
+
+use bsp::entry;
+use hal::clock::GenericClockController;
+use hal::prelude::*;
+use hal::timer::TimerCounter;
+use hal::usb::UsbBus;
+use pac::interrupt;
+use pac::CorePeripherals;
+use pac::Peripherals;
+
+#[entry]
+fn main() -> ! {
+    let mut peripherals = Peripherals::take().unwrap();
+    let mut core = CorePeripherals::take().unwrap();
+    let mut clocks = GenericClockController::with_internal_32kosc(
+        peripherals.GCLK,
+        &mut peripherals.PM,
+        &mut peripherals.SYSCTRL,
+        &mut peripherals.NVMCTRL,
+    );
+    let pins = bsp::Pins::new(peripherals.PORT).split();
+
+    let gclk0 = clocks.gclk0();
+    let timer_clock = clocks.tc4_tc5(&gclk0).unwrap();
+    let mut timer = TimerCounter::tc4_(&timer_clock, peripherals.TC4, &mut peripherals.PM);
+    timer.start(1u32.hz());
+
+    let bus_allocator = unsafe {
+        USB_ALLOCATOR = Some(
+            pins.usb
+                .init(peripherals.USB, &mut clocks, &mut peripherals.PM),
+        );
+        USB_ALLOCATOR.as_ref().unwrap()
+    };
+
+    unsafe {
+        USB_SERIAL = Some(SerialPort::new(&bus_allocator));
+        USB_BUS = Some(
+            UsbDeviceBuilder::new(&bus_allocator, UsbVidPid(0x16c0, 0x27dd))
+                .manufacturer("Fake company")
+                .product("Serial port")
+                .serial_number("TEST")
+                .device_class(USB_CLASS_CDC)
+                .build(),
+        );
+    }
+
+    unsafe {
+        core.NVIC.set_priority(interrupt::USB, 1);
+        NVIC::unmask(interrupt::USB);
+    }
+
+    loop {
+        block!(timer.wait()).ok();
+        unsafe {
+            USB_SERIAL.as_mut().map(|serial| {
+                serial.write("tick!\r\n".as_bytes()).ok();
+            });
+        }
```

Comment:
I haven't actually looked at USB examples before now. Did you write this? Or is this copied from another example?

This seems like UB waiting to happen, because you access `USB_SERIAL` with an `&mut` reference from both `main` and from `poll_usb`. If this were in RTIC, it would require a `Mutex`.

If all the examples are like this, then we can merge it for now, but we should try to get rid of this, I think.

Reply:
Looks like the metro_m0 and feather_m0 examples are different and don't have this problem.

If you want to do something like this, I would **highly** recommend using RTIC. `static mut` variables are *notoriously* difficult to implement soundly. Why try to do it yourself when you can let a well-reviewed code-gen tool do it for you? It's also much easier to understand and gets rid of a lot of the boilerplate you have here.

Reply:
They're not like this, I added this for something for the main loop to do. It's most definitely UB, I was going to ask if anyone knew a nice ergonomic way of doing what I'm doing here. I was considering just wrapping this `free`. I can just drop this have throw a `wfi` in here instead though if you'd prefer.